drivers/usb/bc12/bc12_pi3usb9201.c

```diff
@@ -0,0 +1,383 @@
+	 */
+	pi3usb9201_get_status(dev, NULL, NULL);
+
+	/* Put pi3usb9201 into client mode */
+	rv = pi3usb9201_set_mode(dev, PI3USB9201_CLIENT_MODE);
+	if (rv < 0) {
+		return rv;
+	}
+
+	/* Have pi3usb9201 start bc1.2 detection */
+	rv = pi3usb9201_bc12_detect_ctrl(dev, true);
+	if (rv < 0) {
+		return rv;
+	}
+
+	/* Enable interrupt to wake task when detection completes */
+	return pi3usb9201_interrupt_enable(dev, true);
+}
+
+static void pi3usb9201_client_detect_finish(const struct device *dev, int client_status)
+{
+	struct bc12_partner_state new_chg;
+	int bit_pos;
+
+	/* Set charge voltage to 5V */
+	new_chg.voltage = BC12_CHARGER_VOLTAGE_MV;
+
+	/*
+	 * Find set bit position. Note that this function is only called if a
+	 * bit was set in client_status, so bit_pos won't be negative.
+	 */
+	bit_pos = __builtin_ffs(client_status) - 1;
+
+	new_chg.current = bc12_chg_limits[bit_pos].current_limit;
+	new_chg.type = bc12_chg_limits[bit_pos].partner_type;
+
+	LOG_DBG("client_sts = 0x%x, current = %d mA, type = %d", client_status, new_chg.current,
+		new_chg.type);
+
+	/* bc1.2 is complete and start bit does not auto clear */
+	pi3usb9201_bc12_detect_ctrl(dev, false);
+	/* Inform charge manager of new supplier type and current limit */
+	pi3usb9201_update_charging_partner(dev, &new_chg);
+}
+
+static int pi3usb9201_disconnect(const struct device *dev)
+{
+	/* Put pi3usb9201 into its power down mode */
+	pi3usb9201_set_mode(dev, PI3USB9201_POWER_DOWN);
+	/* The start bc1.2 bit does not auto clear */
+	pi3usb9201_bc12_detect_ctrl(dev, false);
+	/* Mask interrupts unitl next bc1.2 detection event */
+	pi3usb9201_interrupt_enable(dev, false);
+
+	/*
+	 * Let the application know there's no more charge available for the
+	 * supplier type that was most recently detected.
+	 */
+	pi3usb9201_update_charging_partner(dev, NULL);
+
+	return 0;
+}
+
+static int pi3usb9201_set_portable_device(const struct device *dev)
+{
+	/* Disable interrupts during mode change */
+	pi3usb9201_interrupt_enable(dev, false);
+
+	if (pi3usb9201_client_detect_start(dev)) {
+		struct bc12_partner_state new_result;
+
+		/*
+		 * VBUS is present, but starting bc1.2 detection failed
+		 * for some reason. Set the partner type to unknown to disable
+		 * USB data signals and limit current draw to 100 mA.
+		 */
+		new_result.voltage = BC12_CHARGER_VOLTAGE_MV;
+		new_result.current = BC12_CHARGER_MIN_CURR_MA;
+		new_result.type = BC12_TYPE_UNKNOWN;
+		/* Save supplier type and notify callbacks */
+		pi3usb9201_update_charging_partner(dev, &new_result);
+		LOG_ERR("bc1.2 detection failed, using defaults");
+
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void pi3usb9201_isr_work(struct k_work *item)
+{
+	struct pi3usb9201_data *pi3usb9201_data = CONTAINER_OF(item, struct pi3usb9201_data, work);
+	const struct device *dev = pi3usb9201_data->dev;
+	uint8_t client;
+	uint8_t host;
+	int rv;
+
+	rv = pi3usb9201_get_status(dev, &client, &host);
+	if (rv) {
```

Comment:
One more ;)

Reply:
I appreciate your detailed review.  Do you know the specific MISRA rule for this pattern?  I didn't find anything listed under Zephyr's [coding guidelines](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html) that seemed to match. I want to open a github issue to extend the `guidelines_check.py` script to test for compliance with this rule.

Reply:
Here you go https://pvs-studio.com/en/docs/warnings/v2584/.

Reply:
Thanks, it turns out that there is already a [script](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/coccinelle/boolean.cocci) available to verify MISRA 14.4 compliance, but it got disabled.  See https://github.com/zephyrproject-rtos/zephyr/issues/53205 for tracking on this issue.